### PR TITLE
New pointer design

### LINKED ISSUE #196: Pointer/Cursor needs design improvements
REPRODUCE: Home/Everywhere
CURRENT BEHAVIOR: Pointer/cursor needs design improvements
EXPECTATION: Pointer should match rest of design/brand direction.
RECOMMENDATION: Design TBD 

@thenadj @avrignaud (for tracking)

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #196 #156 Added new pointer design

@thenadj I've shared a build with you by email so you can compare between the different pointer colors as requested in the issue. Clicking on the env will loop through the different pointer colors.

### CODE REVIEW

**Comment by MortimerGoro** [comment]
New pointer looks good!

I think that adding a black border to the pointer will be nice. It will make it more visible specially on white backgrounds (e.g. google search)

**Comment by keianhzo** [comment]
Added black rim and reduced tray surface so pointer shows in a more accurate area.

**Comment by bluemarvin** [comment]
On WaveVR the controller is rendered but the pointer transform is not set.

File: app/src/main/cpp/BrowserWorld.cpp [hunk 1]
```diff
@@ -472,6 +507,7 @@ BrowserWorld::InitializeJava(JNIEnv* aEnv, jobject& aActivity, jobject& aAssetMa
     m.controllers->InitializePointer();
     m.loadingAnimation->LoadModels(m.loader);
     m.rootOpaque->AddNode(m.controllers->GetRoot());
+    m.rootOpaqueParent->AddNode(m.controllers->GetPointer());
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply app/src/main/cpp/BrowserWorld.cpp]: controllers->GetPointer() is getting the pointer group not the controller group as I don't want lightning to be applied to the pointer.

File: app/src/main/cpp/Controller.h [hunk 2]
```diff
@@ -42,6 +42,7 @@ struct Controller {
   uint32_t numAxes;
   bool leftHanded;
   bool inDeadZone;
+  vrb::TransformPtr pointerTransform;
```
[the remark this answers is not in the document]
Comment (keianhzo) [reply app/src/main/cpp/Controller.h]: I separated the pointer group as the controller group it's being added to the rootOpaque and if I make it a child of the controller group I can't add it to the rootOpaqueParent.

File: app/src/main/cpp/BrowserWorld.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bluemarvin) [reply app/src/main/cpp/BrowserWorld.cpp]: Sorry, misread the code.

File: app/src/main/cpp/Controller.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (bluemarvin) [reply app/src/main/cpp/Controller.h]: Did you try just setting the ambient color in the material instead? Is the ambient component of the light source not set bright enough?

Comment (bluemarvin) [inline app/src/main/cpp/Controller.h]: You forgot to add this to the Controller::operator=() function which is why it is broken on both Daydream and Wave. I'm not sure why it works on Go.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 3]
```diff
@@ -165,15 +172,20 @@ ControllerContainer::CreateController(const int32_t aControllerIndex, const int3
...
-        controller.transform->AddNode(m.pointerModel);
+        controller.pointerTransform->AddNode(m.pointerModel);
       }
       if (m.root) {
         m.root->AddNode(controller.transform);
         m.root->ToggleChild(*controller.transform, false);
       }
+      if (m.pointerRoot) {
+        m.pointerRoot->AddNode(controller.pointerTransform);
+        m.pointerRoot->ToggleChild(*m.pointerModel, false);
```
Comment (bluemarvin) [inline app/src/main/cpp/ControllerContainer.cpp]: This should be `m.pointerRoot->ToggleChild(*controller.pointerTransform, false);`

This why there is a visible point at the users feet.

**Comment by bluemarvin** [comment]
Okay, figure out why it worked on Go and not Daydream or Wave. Daydream and Wave both report two controllers which causes the the controller std::vector table to get resized. This causes the copy constructor to be called which uses the operator=() function. This causes the pointerTransform to get lost.

**Comment by bluemarvin** [comment]
The daydream controller not showing up is a different issue, still investigating.

**Comment by bluemarvin** [comment]
The Daydream controller was moved to app/src/svr/assets/ where is should have been moved to app/src/googlevr/assets/ so it is missing for Daydream builds.

**Comment by bluemarvin** [comment]
I'm not sure this addresses #156. In order to address #156 the controller needs to be rendered after the dimmer blitter is run. We either need to switch from using the dimmer blitter to the RenderState tint attribute and not set the tint for the controller or, we need to render the controller after the UI has been rendered.

**Comment by keianhzo** [comment]
Thanks for the thorough review @bluemarvin ! 

I thought the only way of getting a shade less material was adding the node to rootOpaqueParent but I've seen that there is a SetLightsEnabled so I've used that and reverted all the pointer handling code.

Also have added another group for non blittered elements to hold the controller.

**Comment by keianhzo** [comment]
@thenadj has approved the design so whenever code looks good you can merge.

File: app/src/main/cpp/BrowserWorld.cpp [hunk 4]
```diff
@@ -138,12 +142,14 @@ struct BrowserWorld::State {
   GroupPtr rootOpaqueParent;
   GroupPtr rootOpaque;
   GroupPtr rootTransparent;
+  GroupPtr rootNoBlitter;
```
Comment (bluemarvin) [inline app/src/main/cpp/BrowserWorld.cpp]: Please rename to rootController. I'll file a follow up to rename the others to rootEnv and rootUi. I don't think we actually need the rootOpaqueParent.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 5]
```diff
@@ -74,8 +74,8 @@ ControllerContainer::InitializePointer() {
   }
   CreationContextPtr create = m.context.lock();
   VertexArrayPtr array = VertexArray::Create(create);
-  const float kLength = -5.0f;
-  const float kHeight = 0.0008f;
+  const float kLength = -1.0f;
+  const float kHeight = 0.005f;
 
   array->AppendVertex(Vector(-kHeight, -kHeight, 0.0f)); // Bottom left
   array->AppendVertex(Vector(kHeight, -kHeight, 0.0f)); // Bottom right
```
Comment (bluemarvin) [inline app/src/main/cpp/ControllerContainer.cpp]: I think the indices are wrong from when the pointer was first created. I'll play around with them and see I can get the triangle surface normals going the correct direction,

**Comment by bluemarvin** [comment]
I think this is the correct vertices indices:

```diff
diff --git a/app/src/main/cpp/ControllerContainer.cpp b/app/src/main/cpp/ControllerContainer.cpp
index 490e7ff..1b7bf1e 100644
--- a/app/src/main/cpp/ControllerContainer.cpp
+++ b/app/src/main/cpp/ControllerContainer.cpp
@@ -100,26 +100,26 @@ ControllerContainer::InitializePointer() {
   std::vector<int> index;
   std::vector<int> uvIndex;
 
-  index.push_back(1);
   index.push_back(2);
+  index.push_back(1);
   index.push_back(5);
   geometry->AddFace(index, uvIndex, index);
 
   index.clear();
-  index.push_back(2);
   index.push_back(3);
+  index.push_back(2);
   index.push_back(5);
   geometry->AddFace(index, uvIndex, index);
 
   index.clear();
-  index.push_back(3);
   index.push_back(4);
+  index.push_back(3);
   index.push_back(5);
   geometry->AddFace(index, uvIndex, index);
 
   index.clear();
-  index.push_back(4);
   index.push_back(1);
+  index.push_back(4);
   index.push_back(5);
   geometry->AddFace(index, uvIndex, index);
```

**Comment by bluemarvin** [comment]
@keianhzo I think if you fix the vertices indices and the button click to change color this will be ready to land. root node refactor can be a follow up.

**Pull request merged**